src/etl/ReportingETL.cpp

```diff
@@ -933,38 +937,55 @@ ReportingETL::loadCacheAsync(uint32_t seq)
-            while (true)
-            {
-                auto res = Backend::synchronousAndRetryOnTimeout(
-                    [this, seq, &cursor](auto yield) {
-                        return backend_->fetchLedgerPage(
-                            cursor, seq, 4096, yield);
-                    });
-                backend_->cache().update(res.objects, seq, true);
-                if (!res.cursor || (end && *(res.cursor) > *end))
-                    break;
-                BOOST_LOG_TRIVIAL(debug)
-                    << "Loading cache. cache size = "
-                    << backend_->cache().size()
-                    << " - cursor = " << ripple::strHex(res.cursor.value());
-                cursor = std::move(res.cursor);
-            }
-            if (--(*numRemaining) == 0)
-            {
-                BOOST_LOG_TRIVIAL(info) << "Finished loading cache";
-                backend_->cache().setFull();
-                delete numRemaining;
-            }
-            else
-            {
-                BOOST_LOG_TRIVIAL(info)
-                    << "Finished a cursor. num remaining = " << *numRemaining;
-            }
-        }};
-        t.detach();
+        cacheDownloadStrands_.emplace_back(ioContext_);
+        boost::asio::spawn(
+            cacheDownloadStrands_.back(),
+            [this, seq, start, end, numRemaining](
+                boost::asio::yield_context yield) {
+                std::optional<ripple::uint256> cursor = start;
+                while (true)
+                {
+                    auto res =
+                        Backend::retryOnTimeout([this, seq, &cursor, &yield]() {
+                            return backend_->fetchLedgerPage(
+                                cursor, seq, 256, yield);
+                        });
+                    backend_->cache().update(res.objects, seq, true);
+                    if (!res.cursor || (end && *(res.cursor) > *end))
+                        break;
+                    BOOST_LOG_TRIVIAL(debug)
+                        << "Loading cache. cache size = "
+                        << backend_->cache().size()
+                        << " - cursor = " << ripple::strHex(res.cursor.value());
+                    cursor = std::move(res.cursor);
+                }
+                if (--(*numRemaining) == 0)
+                {
+                    BOOST_LOG_TRIVIAL(info)
+                        << "Finished loading cache. cache size = "
+                        << backend_->cache().size();
+                    backend_->cache().setFull();
+                    delete numRemaining;
+                }
+                else
+                {
+                    BOOST_LOG_TRIVIAL(info)
+                        << "Finished a cursor. num remaining = "
+                        << *numRemaining;
+                }
+            });
     }
+    // If loading synchronously, poll cache until full
+    while (cacheLoadStyle_ == CacheLoadStyle::SYNC &&
```

Comment:
would it be better to wait on a `condition_variable` checking if cache is full?

Reply:
I thought about that, and I think this loop is probably much simpler